Add fallback value for enums

Luc1412 opened the pull request:
## Summary

It's not unlikely that Epic may add new stuff or messes things up. This could result in the library completely breaking for some endpoints. To prevent this fallback values need to be added.

To solve this issue, enums are parsed with `try_enum`. This will create a fallback unknown enum value in such case.

## TODO:
- [x] Write tests for Enum
- [x] Properly credit Discord.py and Danny, hence code is being used from discord.py repo with MIT license
- [x] Properly apply `try_enum` to Enums that are constructed from a classmethod (`CosmeticCompatibleMode`, `ProductTag`)
- [x] Apply fallback values to other customs classes including `TileSize`

## Checklist

<!-- Put an x inside [ ] to check it, like so: [x] -->

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

Luc1412 committed:
Add try_enum to create unknown value

Luc1412 committed:
Add try_enum to create unknown value

Luc1412 committed:
Fix imports for Enum

Luc1412 committed:
isort, black

Luc1412 committed:
Add copyright info within enums.py

Luc1412 committed:
Add try_enum to enum classmethods

Luc1412 committed:
Add fallback handling for TileSize

Luc1412 committed:
Make the enum name of unknown capitalized

Luc1412 committed:
Add first tests for new enum

Luc1412 committed:
Fix CreatorCodeStatus missing the lower()

Luc1412 committed:
Fix enum test

Luc1412 committed:
Merge branch 'master' into enum-unknown

Luc1412 committed:
Fix enum test

Luc1412 commented on tests/test_enum.py:
The tests for enum are pretty basic. Do we might need to extend it?

trevorflahardy replied:
Absolutely yes, all the defined dunders should be tested as well as a comparison test with `enum.Enum` to guarantee very similar functionality and and interface. I'm not going to note anything on the upcoming review on this file and refer to this conversation instead.

trevorflahardy replied:
Absolutely yes. The tests should ensure that,
1. All the defined functionality on the new `Enum` class works as expected, and
2. The interface that `Enum` provides is the same, or very close to, `enum.Enum` for the user

Luc1412 replied:
I extended the tests. What do you think about it?

Luc1412 commented on fortnite_api/shop.py:
```diff
@@ -123,18 +125,13 @@ def from_value(cls: Type[Self], value: str, /) -> Self:
-            raise ValueError(f"Invalid tile size: {value!r}")
+            # Epic may change the format or messes up the value.
+            # In one such case, this value was "ERROR BAD SIZE".
+            raise cls(width=1, height=1, internal=value)
```
I choose 0x0 as a fallback value for TileSize, since in case of `ERROR BAD SIZE` the item isn't rendered. Not sure if there is a better option.

trevorflahardy replied:
I think this should actually be `-1` to indicate a failed parsing. What do you think?

Luc1412 replied:
`-1` is common as an error indicator, but am not sure. 0x0 would be also a value that shouldn't be returned by the API.

Luc1412 committed:
Merge branch 'master' into enum-unknown

Luc1412 committed:
Change default width and height of TileSize to 0

Luc1412 committed:
Add changelog

trevorflahardy commented on docs/changelog.rst:
```diff
@@ -14,6 +14,12 @@ New Features
 
+Miscellaneous
+~~~~~~~~~~~~~
+- Add safeguards against Fortnite changes or messed up values within the API.
```
Could be made more clear here
```suggestion
- Add safeguards against Epic Games' API changing or providing invalid values in API responses.
```

trevorflahardy replied:
I'd also note that all "Enum-like" classes in the breaking changes section are not actually `enum.Enum`s anymore, in case one of our users is using an instance check.

Luc1412 replied:
~Shouldn't instance checks still work, since we just subclassing Enum? But yeah, if you expect a specific type it could break.~ Nvm. completely forgot that the Enum import was just for type checking. Gonna add that to the changelog.

Luc1412 replied:
I updated the changelog. Feel free to suggest changes in wording.

trevorflahardy commented on fortnite_api/enums.py:
```diff
@@ -47,7 +49,133 @@
+E = TypeVar('E', bound='Enum')
+
+
+def _create_value_cls(name: str, comparable: bool):
```
Missing return annotation here. It may be worth it to define a subclass of `typing.NamedTuple` instead of the constructor so the return value is correct (?)

trevorflahardy replied:
As a side note, you have a pyright error in the function as well:
```python
"namedtuple" provides no types for tuple entries; use "NamedTuple" instead
```

Addressing the return annotation would also alleviate the issue

Luc1412 replied:
I implemented a possible solution. What are your thoughts?

Luc1412 replied:
Seems like it messes a lot up: https://github.com/Fortnite-API/py-wrapper/actions/runs/11980161029/job/33403864882

trevorflahardy replied:
The approach is good, adjusted in [cf883e6](https://github.com/Fortnite-API/py-wrapper/pull/33/commits/cf883e612375b66fbe68c6cc755fa2eab41906af) 👍

Luc1412 replied:
Moving _EnumValue out of the function, seem to break the Enum. When checking the tests, Moving _EnumValue out of the function, seem to break the Enum. When checking the tests, the ProductTag Enum still is the DummyEnum from the enums tests.

trevorflahardy commented on fortnite_api/enums.py:
```diff
@@ -47,7 +49,133 @@
+        cls.__ge__ = lambda self, other: isinstance(other, self.__class__) and self.value >= other.value  # type: ignore
+        cls.__lt__ = lambda self, other: isinstance(other, self.__class__) and self.value < other.value  # type: ignore
+        cls.__gt__ = lambda self, other: isinstance(other, self.__class__) and self.value > other.value  # type: ignore
+    return cls
```
I believe black should have done this for you - ensure you run with a black format. If not, then that's atypical for the black formatter. Re-run to be safe :)
```suggestion
    
    return cls
```

trevorflahardy commented on fortnite_api/enums.py:
```diff
@@ -47,7 +49,133 @@
+    return cls
+
+
+def _is_descriptor(obj):
```
Missing annotations here,

```suggestion
def _is_descriptor(obj: Type[object]) -> bool:
```

As a side note, `typing.TypeIs` may fit well here alongside an internally defined `Protocol` in the `TYPE_CHECKING` block,
```python
if TYPE_CHECKING:
    class _SetterObj(typing.Protocol):
        # Defined methods
        def __get__(...): ...

        # ...

def predicate(obj: Type[object]) -> TypeGuard[_SetterObject]: ...
```
What do you think?

trevorflahardy commented on fortnite_api/enums.py:
```diff
@@ -47,7 +49,133 @@
+
+class EnumMeta(type):
+    if TYPE_CHECKING:
+        __name__: ClassVar[str]
```
I don't think this is needed. `__name__` here actually overrides the `__name__` attribute on the `"type"` class (may need to be defined if moved, though - see conversation below for more)

```suggestion
```

trevorflahardy commented on fortnite_api/enums.py:
```diff
@@ -47,7 +49,133 @@
+        *,
+        comparable: bool = False,
+    ) -> EnumMeta:
+        value_mapping = {}
+        member_mapping = {}
+        member_names = []
+
```
Missing definition annotations, causing every reference to these to be unknown.

```suggestion
        value_mapping: Dict[OldValue, NewValue] = {}
        member_mapping: Dict[str, NewValue] = {}
        member_names: List[str] = []

```
Where `OldValue` and `NewValue` are typealiases that simply point to `Any`

trevorflahardy replied:
Is pyright picking up the `pyproject.toml` on your local development environment? This raised multiple type errors for me.

Luc1412 replied:
I have to admit that Pyright is currently disabled in my IDE, hence the plugin isn't updated for the latest PyCharm version 🥲

trevorflahardy commented on fortnite_api/enums.py:
```diff
@@ -454,4 +582,22 @@ def _from_str(cls: Type[Self], string: str) -> Self:
+        return try_enum(cls, trimmed.lower())
+
+
+def create_unknown_value(cls: Type[E], val: Any) -> E:
+    value_cls = cls._enum_value_cls_  # type: ignore # This is narrowed below
+    name = f'UNKNOWN_{val}'
+    return value_cls(name=name, value=val)
+
+
+def try_enum(cls: Type[E], val: Any) -> E:
+    """A function that tries to turn the value into enum ``cls``.
+
+    If it fails it returns a proxy invalid value instead.
+    """
+
+    try:
+        return cls._enum_value_map_[val]  # type: ignore # All errors are caught below
+    except (KeyError, TypeError, AttributeError):
+        return create_unknown_value(cls, val)
```
All the type ignores can be removed here by doing the following,
- Remove the type checking patch at the top of this file (`if TYPE_CHECKING: from enum import Enum`) such that the custom `Enum` class is defined at all times,
- Correct the `_enum_value_map_` and related `ClassVars` by moving them to the `Enum` class definition, and
- Add the missing `_enum_value_cls` ClassVar on the `Enum` class definition.

This points `TypeVar('E', bound='Enum')` to the right type and, in doing so, doesn't convince the user that some enum in the library is `enum.Enum` when it isn't.

Let me know if this logic doesn't pan out, but I tested it locally and the type errors went away. Currently the typing indicates that,
```python
class MyEnum(Enum):
    FOO = 1

print(MyEnum.__class__._enum_value_map_)
```
is valid syntax, when in actuality it is `MyEnum._enum_value_map_`.

Luc1412 replied:
Couldn't figure out a solution based on your description that doesn't cause more type errors. Feel free to implement your solution right within this PR.

trevorflahardy replied:
Appended in cf883e6 👍

trevorflahardy commented on fortnite_api/enums.py:
```diff
@@ -47,7 +49,133 @@
+
+
+def _create_value_cls(name: str, comparable: bool):
+    # All the type ignores here are due to the type checker being unable to recognise
+    # Runtime type creation without exploding.
+    cls = namedtuple('_EnumValue_' + name, 'name value')
```
I'd make this a bit more clear haha,

```suggestion
    # Pyright cannot statically recognize the runtime type creation. All of the following
    # type ignores in this function are a result of this.
    cls = namedtuple('_EnumValue_' + name, 'name value')
```

trevorflahardy commented on docs/changelog.rst:
```diff
@@ -14,6 +14,12 @@ New Features
+Miscellaneous
+~~~~~~~~~~~~~
+- Add safeguards against Fortnite changes or messed up values within the API.
+    - All enums now can handle unknown values. This means that if the API returns a value that is not in the enum, it will be stored as an attribute on the enum object.
```
```suggestion
    - All enums now can handle unknown values via an internally defined "Enum-like" class. This means that if the API returns a value that is not in the enum, it will be stored as an attribute on the enum object.
```

Luc1412 committed:
Apply suggestions from code review

Co-authored-by: Trevor Flahardy <75498301+trevorflahardy@users.noreply.github.com>

Luc1412 committed:
Merge branch 'master' into enum-unknown

Luc1412 committed:
black

Luc1412 committed:
Merge branch 'master' into enum-unknown

Luc1412 committed:
Add missing OldValue and NewValue type aliases

Luc1412 committed:
Merge remote-tracking branch 'origin/enum-unknown' into enum-unknown

Luc1412 committed:
Rework the value cls creation by defining a NamedTuple as a subclass of typing.NamedTuple. This also allows the removal of #type: ignore's

Luc1412 committed:
Revert "Rework the value cls creation by defining a NamedTuple as a subclass of typing.NamedTuple. This also allows the removal of #type: ignore's"

This reverts commit 907e6724573535993901f6f34c89ef8780e9bfce.

Luc1412 committed:
Rework the value cls creation by defining a NamedTuple as a subclass of typing.NamedTuple.

Luc1412 committed:
Extend tests for custom Enum-like type

Luc1412 committed:
Add missing pytest import

Luc1412 committed:
Update changelog for breaking enum change

trevorflahardy commented on docs/changelog.rst:
```diff
@@ -14,12 +14,19 @@ Breaking Changes
 ~~~~~~~~~~~~~~~~
 - ``CreatorCode.status`` and ``CreatorCode.disabled`` have been removed, since they where never returned by this endpoint. Disabled creator codes always raise :class:`fortnite_api.NotFound` when trying to fetch them.
 - ``CreatorCode.verified`` has been removed, since it isn't used within the affiliate system. It always returns ``False``.
+- All enums now use an internal "Enum-like" class to handle unknown values, instead of the built-in ``enum.Enum``. This potentially breaks type checks, but it should not break core functionality.
```
```suggestion
- All enums now use an internal "Enum-like" class to handle unknown values, instead of the built-in :class:`py.enum.Enum`. This potentially breaks type checks, but does not break core functionality or change the enum interface; you can use them the same.
```

trevorflahardy committed:
Update enums.py for all typing changes required to conform to Pyright strict mode.

trevorflahardy committed:
Widen the typing on try_value

trevorflahardy commented on docs/changelog.rst:
```diff
@@ -14,12 +14,19 @@ Breaking Changes
+Miscellaneous
+~~~~~~~~~~~~~
+- Add safeguards against Epic Games' API changing or providing invalid values in API responses.
+    - All enums now can handle unknown values via an internally defined "Enum-like" class. This means that if the API returns a value that is not in the enum, it will be stored as an attribute on the enum object.
```
```suggestion
    - All enums now can handle unknown values via an internally defined "Enum-like" class. If the API returns a value not in the enum, it will be stored as an attribute on the enum object. The interface for using this class is the same as using :class:`py.enum.Enum`.
```

Luc1412 committed:
Apply suggestions from code review

Co-authored-by: Trevor Flahardy <75498301+trevorflahardy@users.noreply.github.com>

trevorflahardy committed:
- Update enums.py to revert issue created by moving _EnumValue out of a localized function.
- Update enums.py to address some Python 3.9 typing styles due to the recent merged PR that bumps lib to >=3.9

Luc1412 committed:
update branch

trevorflahardy committed:
Black isort on files

Luc1412 committed:
Merge remote-tracking branch 'origin/enum-unknown' into enum-unknown

Luc1412 committed:
Merge branch 'master' into enum-unknown

Luc1412 committed:
Upgrade imports for python 3.9

Luc1412 committed:
Fix Python enum reference

Luc1412 committed:
Update TileSize fallback value to -1

Luc1412 committed:
Cleanup changelog formatting

Luc1412 committed:
Update changelog regarding TileSize
